docs: add versioning policy (#3332)
DataDog/dd-trace-py: docs/versioning.rst

**********
Versioning
**********

Release support
===============

.. list-table::
   :header-rows: 1

   * - Release
     - :ref:`Support level<versioning_support_levels>`

>      - Minimum Trace Agent

Author: ```suggestion
     - Minimum Datadog Agent
```

   * - 0.x
     - :ref:`Maintenance<versioning_support_maintenace>`

>      -

Reviewer: I think this is needed no? since there are 3 columns?

>      -

Reviewer 2: ah, could be!

   * - 1.x
     - :ref:`General Availability<versioning_support_ga>`
     - 7.28

Versions
========

The non-negative integer components of the **version format** (``v<MAJOR>.<MINOR>.<PATCH>``) are incremented by the criteria:

MAJOR
    Incompatible changes to the public :ref:`interface<versioning_interfaces>`

    Removing support for a :ref:`runtime<versioning_supported_runtimes>`.

MINOR
    Backwards compatible changes to the public :ref:`interface<versioning_interfaces>`

    Any backwards compatible or incompatible changes to the internal :ref:`interface<versioning_interfaces>`

>     Adding a support for :ref:`runtime<versioning_supported_runtimes>`

Reviewer: ```suggestion
    Adding support for a :ref:`runtime<versioning_supported_runtimes>`
```

?

PATCH
    Bug fixes that are backwards compatible

    Security fixes that are backwards compatible

.. _versioning_interfaces:

Interfaces
==========

The definition of the **internal** interface:

>     The ``ddtrace.internal`` package is internal

Reviewer 2: ```suggestion
    The ``ddtrace.internal`` module is internal
```

Author: I think the `package` term is more apt here as it bundles modules together:

https://docs.python.org/3/tutorial/modules.html#packages

>     The ``ddtrace.vendor`` package is internal

Reviewer 2: ```suggestion
    The ``ddtrace.vendor`` module is internal
```

    Any module, function, class or attribute that is prefixed with a single underscore is internal

>     Any package, module, function, class or attribute that is contained within an internal package is internal

Reviewer 2: ```suggestion
    Any module, function, class or attribute that is contained within an internal module is internal
```

The definition of the **public** interface:

>     Any package, module, function, class or attribute that is not internal

Reviewer: ```suggestion
    Any module, function, class or attribute that is not internal
```

?

.. _versioning_supported_runtimes:

Supported runtimes
==================

.. list-table::
   :header-rows: 1

   * - OS
     - CPU
     - Runtime
     - Runtime version
     - Supported Release
   * - Linux
     - x86-64, i686, AArch64
     - CPython

>      - 3.10, 3.9, 3.8, 3.7, 3.6, 3.5, 2.7

Reviewer: ```suggestion
     - 2.7, 3.5-3.10
```

**********
Versioning
**********

Current Release Stages
======================

.. list-table:: Releases
   :widths: 20 30 30 20
   :header-rows: 1

   * - Release
     - Runtimes
     - Minimum Trace Agent
     - Stage
   * - 0.x
     - CPython: 2.7, 3.5–3.10
     - 
     - MAINTENANCE [2022-03-01]
   * - 1.x
     - CPython: 2.7, 3.5+
     - 7.28
     - CURRENT

Runtimes
========

> .. list-table:: Supported runtime versions

Reviewer: this table isn't rendering for me in GitHub preview, but the one above is, I can't tell what is wrong with it (if anything is)

Reviewer: same with OS

   :widths: 60 40
   :header-rows: 1

   * - Runtime
     - Runtime version
   * - CPython 
     - 3.10
     - 3.9
     - 3.8
     - 3.7
     - 3.6

>      - 2.7

Reviewer 3: Double-checking that we're not missing 3.5 here

   
.. list-table:: Supported operating system
   :widths: 60 40
   :header-rows: 1

   * - Operating system 
     - CPU architecture
     - Tested in CI
   * - Linux
     - x86-64
     - :white_check_mark:
   * - Linux
     - i686
   * - Linux
     - AArch64
   * - MacOS
     - Intel
   * - MacOS
     - Apple Silicon
   * - Windows
     - 64bit
   * - Windows
     - 32bit

Interfaces
==========

The definition of the **internal** interface:
    The ``ddtrace.internal`` package is internal

    The ``ddtrace.vendor`` package is internal

    Any module, function, class or attribute that is prefixed with a single underscore is internal

>     Any package, module, function, class or attribute that is contained within an internal namespace is also internal

Reviewer 3: Maybe worth clarifying what we mean by "internal namespace" here?

Author: Dropped the term "namespace" and instead stuck with "package" as the Pythonic term.

The definition of the **public** interface:
    Any package, module, function, class or attribute that is not internal

Versions
========

The non-negative integer components of the **version format** (``v<MAJOR>.<MINOR>.<PATCH>``) are incremented by the criteria:

> MAJOR

Reviewer: since we define OS support above we should add that here as well for dropping/adding support.

    Incompatible changes to the public interface

>     Dropping support for a Python runtime

Reviewer: ```suggestion
    Dropping support for a Python runtime version
```

? do we need to clarify this? does "Runtime version" mean "CPython" or "CPython 3.7" ?

Reviewer: ```suggestion
    Removing support for a Python runtime
```

?

MINOR
    Backwards compatible changes to the public interface

>     Any changes to the internal interface

Reviewer: "including backwards-incompatible changes"

wondering if we should be more explicit here compared with "any".

what do you think?

    Adding support for a Python runtime

PATCH

>     Bug fixes that are backwards compatible

Author: Not sure if the hyphenation is necessary here. Is there a good rule for when we hyphenate?

    Security fixes that are backwards compatible

Release stages
==============

The **stage** of each release determines when changes are made and the level of support:

BETA
    Under active development.

    Receives backwards incompatible features.
    
    Not supported for production use.

    Will become the next CURRENT release.

> GENERAL AVAILABILITY [GA]

Reviewer: CURRENT?

Author: I think GA is more specific and consistent with the release cycles of other ddtrace libraries.

**********
Versioning
**********

> Current Release Stages

Reviewer: should we have a link in this section to the definitions below?

or link each status name in the table to the definition below? (footnote?)

======================

.. list-table:: Releases
   :widths: 20 30 30 20
   :header-rows: 1

   * - Release
     - Runtimes
     - Minimum Trace Agent
     - Stage
   * - 0.x
     - CPython: 2.7, 3.5–3.10

>      - 

Reviewer: we should define a minimum, we can pick an old 6.x version or can even just check what the oldest version any customer is currently using and use that. what do you think?

     - MAINTENANCE [2022-03-01]
   * - 1.x
     - CPython: 2.7, 3.5+
     - 7.28
     - CURRENT

Runtimes
========

> .. list-table:: Supported runtime versions

Reviewer: Is there a reason this is centered in the page? do we control that?

<img width="644" alt="image" src="https://user-images.githubusercontent.com/1320353/160618075-855d687d-4f35-48d4-96da-2af1e8de674a.png">

Reviewer: 👍🏻 then ignore, thanks!

   :widths: 60 40
   :header-rows: 1

   * - Runtime

>      - Runtime version

Reviewer: should we include "tested in CI" same as OS?

if we ever decide to stop testing 3.5 but want to keep supporting it... would we ever do that? or maybe just good to be explicit?

Author: I am not totally happy with these tables as is.

We really want a table of `OS x Arch x Runtime` with whether the combination is tested in CI and we distribute wheels. wdyt?

Reviewer: yeah, I agree, that sounds great.

verbose, but we can figure out a concise way to represent it... 🤞🏻

   * - CPython 
     - 3.10
   * -
     - 3.9
   * -
     - 3.8
   * -
     - 3.7
   * -
     - 3.6
   * -
     - 3.5
   * -
     - 2.7

   

> .. list-table:: Supported operating systems

Reviewer: I wonder if we need to merge this + last tables

if we ever add support for pypy maybe we only want to support linux x86-64?

would we ever want to do that?

*****************
Versioning Policy
*****************

> The internal interface of the `ddtrace` library is defined as follows:

Reviewer: ```suggestion
The internal interface of the ``ddtrace`` library is defined as follows:
```

    Any module, function, class or attribute is internal if it is prefixed with a single underscore

>     Any package that is named `internal` is internal

Reviewer: ```suggestion
    Any package that is named ``internal`` is internal
```

Reviewer: Should we just rename `internal` to `_internal`?

We should also call out and underscore prefix `_vendor` as well?

Author: Yep, using the single underscore prefix as the only criteria for what constitutes an internal interface would simplify this section.

>     Any package, module, function, class or attribute that is contained within a namespace that is internal is also internal

Reviewer: ```suggestion
    Any package, module, function, class or attribute that is contained within a namespace that is internal is also internal.
```

Author: Leaving off the sentence punctuation as these term definitions are not sentences and punctation is not used elsewhere in this section.

When a new major version is released it becomes the primary release line, where all new features, bug fixes and security patches land.

> The non-negative integers components of the *vMAJOR.MINOR.PATCH* format are incremented by the following rules:

Reviewer 3: ```suggestion
The non-negative integer components of the *v<MAJOR>.<MINOR>.<PATCH>* format are incremented by the following rules:
```

MAJOR
    API changes incompatible with previous versions

    Functionality changes incompatible with previous versions

>     Dropping support for anything, but not limited to, Python versions, library versions supported in integrations, or Datadog APM features.

Reviewer: > library versions supported in integrations

🤔 I am not sure we figured this part out yet.

Keeping in `MINOR` is good, but maybe we should remove from here.

e.g. if we dropped support for an older version of Django we probably wouldn't do a major version bump (we _might_ want to, but we haven't really defined our process there yet, I'd hate to commit to it without exploring further)

Author: Agree -- I put it here but with that same question.

Author: Can we still commit to dropping support for a Python version only in major versions?

Reviewer: 100% that is definitely a breaking change that requires a major version bump.

MINOR
    API changes that are backwards compatible

    Functionality changes that are backwards compatible

    Adding support for anything, but not limited to, Python versions, library versions supported in integrations, or Datadog APM features.

PATCH
    Bug fixes that are backwards compatible

    Security fixes that are backwards compatible

> .. list-table:: Status of Release Lines

Reviewer: move to the top. the quick/easy to scan thing should be at the top

   :widths: 20 30 30 20
   :header-rows: 1

   * - Release Line
     - Supported Runtimes

>      - Minimum Trace Agent

Reviewer: I wonder if this is something we can detect and report on in the library 🤔

Reviewer: e.g. on startup try to hit the info endpoint, if we cannot or the response is a version too low, we report that the agent needs to be upgraded?

(necessary for 1.0.0 release?)

Author: Let's investigate if this is possible though give users a way to opt out of the guards.

     - Status
   * - 0.x
     - CPython: 2.7, 3.5+
     - -

>      - Maintenance

Reviewer: I was wondering if we should add dates here. The date it went into maintenance?

>    * - 1.x

Reviewer: I wonder if we should flip the order, most recent at the top?

     - CPython: 2.7, 3.5+
     - 7.22
     - Current

>    * - 1.x
>      - CPython: 3.6+
>      - TBD
>      - Future

Reviewer: ```suggestion
```

I added this in the draft doc, but I don't think we should call it out here.

Yes, we will want to drop 2.7 support at some point, but we aren't 100% sure it'll be 2.x.

**********
Versioning
**********

Current Status
==============

.. list-table:: Releases
   :widths: 20 30 30 20
   :header-rows: 1

   * - Release
     - Supported Runtimes
     - Minimum Trace Agent
     - Status
   * - 0.x
     - CPython: 2.7, 3.5+
     - 

>      - MAINTENANCE [2022-02-25]

Reviewer: TODO: Update this date.

(just leaving a comment so we don't forget)

   * - 1.x

>      - CPython: 2.7, 3.5+

Reviewer 4: ```suggestion
     - CPython: 3.6+
```
🙏

Reviewer: Not yet... but I think we can do this with a planned 2.0 release.

My thinking is we can introduce these terms, maintenance definitions for 1.0, so that way when we do drop support for 2.7 people understand what they are going to get/how we are going to continue to support them going forward.

     - 7.28
     - CURRENT

> The **status** of each release determines when changes are made:

Reviewer 4: Do we have a status for "not maintained anymore"?

MAINTENANCE
    No longer under development.

>     Receives bug fixes and security fixes.

Reviewer: ```suggestion
    Supported with bug fixes and security fixes.
```

?

Reviewer: Then can make the same change to CURRENT.

The reason for this is we can then make sure to mark DEVELOPMENT as being "not supported" (not sure exact wording).

what do you think?

CURRENT
    Under active development.

    Receives bug fixes and security fixes.

    Receives backwards compatible features.

DEVELOPMENT
    Under active development.

    Receives backwards incompatible features.

    Will become the next CURRENT release.

Interface
=========

> The definition of the **internal** interface:

Reviewer: We might need to explicitly state that breaking changes can be made to the internal interface at any time without warning or a major version bump.

Might be good to cover this in the "Versions" section, and then move this Interface section to the end?

    The ``ddtrace.internal`` package is internal

    The ``ddtrace.vendor`` package is internal

    Any module, function, class or attribute that is prefixed with a single underscore is internal

    Any package, module, function, class or attribute that is contained within an internal namespace is also internal

The definition of the **public** interface:
    Any package, module, function, class or attribute that is not internal

Versions
========

The non-negative integer components of the **version format** (``v<MAJOR>.<MINOR>.<PATCH>``) are incremented by the criteria:

MAJOR

>     API changes incompatible with previous versions

Reviewer: maybe we should use "public"/"internal" in these definitions to make the "Interface" section ?

> API changes incompatible with previous version's public interface

>     Functionality changes incompatible with previous versions

Reviewer: I wonder if we need to expand on this, or be more specific.

I wonder how integrations fall under this category?

**********
Versioning
**********

Current Status
==============

> When a new major version is released it becomes the current release line, where all new features, bug fixes and security patches land.

Reviewer: do we want this below the table? seems more fitting to have this statement near the definitions of current/development/etc.

do we need a definition of "release line"?

Author: I've opted remove this text and not refer to releases as "release lines". I think "line" is more relevant to when we discuss our git branching, which is beyond the scope of this versioning policy.

.. list-table:: Release Lines
   :widths: 20 30 30 20
   :header-rows: 1

   * - Line
     - Supported Runtimes

>      - Minimum Trace Agent

Reviewer 5: Should we recommend using the latest version of the trace agent? Certain features (ex. telemetry) are only supported in recent trace agent release. Here we could add a `Recommended trace agent` column

Reviewer: that is what the minimum version is meant to help with. I could see us adding a footnote/aside that says "we always recommend using the latest version of the trace agent". but I would want us to say "min version is 7.22 but recommended is 7.35" that sends mixed messages.

     - Status
   * - 0.x
     - CPython: 2.7, 3.5+
     - 
     - MAINTENANCE [2022-02-25]
   * - 1.x
     - CPython: 2.7, 3.5+

>      - 7.22

Reviewer: ```suggestion
     - 7.28
```

This version if almost 1 year old, and is min needed to support stats computation

     - CURRENT

>    * - 2.x
>      - TBD
>      - TBD
>      - DEVELOPMENT

Reviewer: ```suggestion
```

The **status** of each release line determines when changes are made:

MAINTENANCE
    No longer under development.

>     Receives all bug fixes and security fixes.

Reviewer: ```suggestion
    Receives bug fixes and security fixes.
```

    Will not receive any new features.

CURRENT
    Under active development.

>     Receives all backwards compatible new features, bug fixes, and security fixes.

Reviewer: ```suggestion
    Receives new features, and backwards compatible bug and security fixes.
```

? 

new features should always be backwards compatible ("should"?).

DEVELOPMENT

>     Under active development.
> 
>     Receives all incompatible new features.
> 
>     Not yet released or maintained.

Reviewer: ```suggestion
    Under active development.

    Receives all incompatible new features.

    Not yet released or maintained.
    
    Will become the next current version when released.
```

Do we want to note somewhere about how as soon as a new major release line is released, the CURRENT will become MAINTENANCE?

FUTURE

>     Planned but not under active development.

Reviewer: I wonder if we need FUTURE? could be nice to announce a new version is coming and runtime + agent requirements, but then wouldn't it just be under "DEVELOPMENT"?

Interface
=========

> The definition of the **internal** interface:

Reviewer: ```suggestion
The definition of the **private** interface:
```

? probably just splitting hairs here.

Author: I prefer "internal" over "private" as this is the distinction in PEP 8: https://peps.python.org/pep-0008/#public-and-internal-interfaces

    Any module, function, class or attribute is internal if it is prefixed with a single underscore

>     Any package that is named ``internal`` is internal

Reviewer: ```suggestion
    Any module that is named ``internal`` is internal
```

? or do we say "package" because we meant he collection of modules under the name `*.internal` ? 🤔

Author: We would say "package" here. The spec for internal "module" is covered by the first item (files are prefixed with single underscore). The spec for internal "package" is any package directory that is named as such. We could also say any directory with a prefix of single underscore but I do not see that in our current directory structure.

    Any package, module, function, class or attribute that is contained within a namespace that is internal is also internal

The definition of the **public** interface:
    Any package, module, function, class or attribute not in the internal interface

>     Any package, module, function, class or attribute not contained within the ``ddtrace.vendor`` package.

Reviewer: do we just move this up to the internal definition?

then "public interface" just because "everything else"

Author: That makes sense. Any vendor package is also internal.
